File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    /*
    
    */
    
    let set = new Set();
    let max = 0;
    let left = 0;
    for(let i=0; i<s.length; i++){
        while(set.has(s[i])){
            set.delete(s[left]);
            left++;
        }
        set.add(s[i])
        max = Math.max(max, i-left+1)
    }
    return max
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        '''
        set = {c,a}
                       r 
        Input: s = "baca"
                      l
                      
        curLongest              
        maxlongest = 3
        
        Output: 3
        '''
        hset = set()
        left = longest = 0
        for right in range(len(s)):          
            while s[right] in hset:
                hset.remove(s[left])
                left += 1
            
            hset.add(s[right])

                
            longest = max(longest, right-left+1)
        
        return longest
    
    
    